File: go/dotprompt/dotprompt.go
// Copyright 2025 Google LLC
// SPDX-License-Identifier: Apache-2.0

package dotprompt

import (
	"errors"
	"fmt"
	"reflect"
	"regexp"
	"strings"

	"github.com/aymerick/raymond"
)

// PartialResolver is a function to resolve partial names to their content.
type PartialResolver func(partialName string) (string, error)

// DotpromptOptions defines the options for the Dotprompt instance.
type DotpromptOptions struct {
	DefaultModel    string
	ModelConfigs    map[string]any
	Helpers         map[string]any
	Partials        map[string]string
	Tools           map[string]ToolDefinition
	ToolResolver    ToolResolver
	Schemas         map[string]JSONSchema
	SchemaResolver  SchemaResolver
	PartialResolver PartialResolver
}

// Dotprompt is the main struct for the Dotprompt instance.
type Dotprompt struct {
	knownHelpers    map[string]bool
	defaultModel    string
	modelConfigs    map[string]any
	tools           map[string]ToolDefinition
	toolResolver    ToolResolver
	schemas         map[string]JSONSchema
	schemaResolver  SchemaResolver
	partialResolver PartialResolver
	knownPartials   map[string]bool
}

// NewDotprompt creates a new Dotprompt instance with the given options.
func NewDotprompt(options *DotpromptOptions) *Dotprompt {
	dp := &Dotprompt{
		knownHelpers: make(map[string]bool),
	}
	if options != nil {
		dp = &Dotprompt{
			knownHelpers:    make(map[string]bool),
			modelConfigs:    options.ModelConfigs,
			defaultModel:    options.DefaultModel,
			tools:           options.Tools,
			toolResolver:    options.ToolResolver,
			schemas:         options.Schemas,
			schemaResolver:  options.SchemaResolver,
			partialResolver: options.PartialResolver,
			knownPartials:   make(map[string]bool),
		}
	}

	return dp
}

// DefineHelper registers a helper function.
func (dp *Dotprompt) defineHelper(name string, helper any, tpl *raymond.Template) {
	if dp.knownHelpers[name] {
		return
	}
	tpl.RegisterHelper(name, helper)
	dp.knownHelpers[name] = true
}

// DefinePartial registers a partial template.
func (dp *Dotprompt) definePartial(name string, source string, tpl *raymond.Template) {
	if dp.knownPartials[name] {
		return
	}
	tpl.RegisterPartial(name, source)
	dp.knownPartials[name] = true
}

// TODO: Add register helpers
func (dp *Dotprompt) RegisterHelpers(options *DotpromptOptions, tpl *raymond.Template) {
	if options != nil && options.Helpers != nil {
		for key, helper := range options.Helpers {
			dp.defineHelper(key, helper, tpl)
		}
	}
	for name, helper := range templateHelpers {
		dp.defineHelper(name, helper, tpl)
	}
}

func (dp *Dotprompt) RegisterPartials(options *DotpromptOptions, tpl *raymond.Template, template string) error {
	if options.Partials != nil {
		for key, partial := range options.Partials {
			dp.definePartial(key, partial, tpl)
		}
	}
	err := dp.resolvePartials(template, tpl)
	if err != nil {
		return err
	}
	return nil
}

// DefineTool registers a tool definition.
func (dp *Dotprompt) DefineTool(def ToolDefinition) *Dotprompt {
	dp.tools[def.Name] = def
	return dp
}

// Parse parses the source string into a ParsedPrompt.
func (dp *Dotprompt) Parse(source string) (ParsedPrompt, error) {
	return ParseDocument(source)
}

// Render renders the source string with the given data and options.
func (dp *Dotprompt) Render(source string, data *DataArgument, options *PromptMetadata, dotpromptOptions *DotpromptOptions) (RenderedPrompt, error) {
	renderer, err := dp.Compile(source, options, dotpromptOptions)
	if err != nil {
		return RenderedPrompt{}, err
	}
	return renderer(data, options)
}

// Compile compiles the source string into a PromptFunction.
func (dp *Dotprompt) Compile(source string, additionalMetadata *PromptMetadata, dotpromptOptions *DotpromptOptions) (PromptFunction, error) {
	parsedPrompt, err := dp.Parse(source)
	if err != nil {
		return nil, err
	}
	if additionalMetadata != nil {
		parsedPrompt = mergeMetadata(parsedPrompt, additionalMetadata)
	}

	renderTpl, err := raymond.Parse(parsedPrompt.Template)
	if err != nil {
		return nil, err
	}

	// RegisterHelpers()
	dp.RegisterHelpers(dotpromptOptions, renderTpl)
	err = dp.RegisterPartials(dotpromptOptions, renderTpl, parsedPrompt.Template)
	if err != nil {
		return nil, err
	}

	renderFunc := func(data *DataArgument, options *PromptMetadata) (RenderedPrompt, error) {
		mergedMetadata, err := dp.RenderMetadata(parsedPrompt, options)
		if err != nil {
			return RenderedPrompt{}, err
		}

		inputContext := MergeMaps(data.Input, data.Context)
		renderedString, err := renderTpl.Exec(inputContext)

		if err != nil {
			return RenderedPrompt{}, err
		}

		messages, err := ToMessages(renderedString, data)
		if err != nil {
			return RenderedPrompt{}, err
		}
		return RenderedPrompt{
			PromptMetadata: mergedMetadata,
			Messages:       messages,
		}, nil
	}

	return renderFunc, nil
}

// IdentifyPartials identifies partials in the template.
func (d *Dotprompt) identifyPartials(template string) []string {
	// Simplified partial identification logic
	var partials []string
	lines := strings.Split(template, "\n")
	for _, line := range lines {
		re := regexp.MustCompile(`{{>\s*([^}]+)\s*}}`)
		// Find all matches in the template
		matches := re.FindAllStringSubmatch(line, -1)

		for _, match := range matches {
			if len(match) > 1 {
				partialName := strings.TrimSpace(match[1])
				partials = append(partials, partialName)
			}
		}
	}
	return partials
}

// resolvePartials resolves and registers partials in the template.
func (dp *Dotprompt) resolvePartials(template string, tpl *raymond.Template) error {
	if dp.partialResolver == nil {
		return nil
	}

	partials := dp.identifyPartials(template)
	for _, partial := range partials {
		if _, exists := dp.knownPartials[partial]; !exists {
			content, err := dp.partialResolver(partial)
			if err != nil {
				return err
			}
			if content != "" {
				dp.definePartial(partial, content, tpl)
				err = dp.resolvePartials(content, tpl)
				if err != nil {
					return err
				}
			}
		}
	}
	return nil
}

// mergeMetadata merges additional metadata into the parsed prompt.
func mergeMetadata(parsedPrompt ParsedPrompt, additionalMetadata *PromptMetadata) ParsedPrompt {
	if additionalMetadata != nil {
		if additionalMetadata.Model != "" {
			parsedPrompt.PromptMetadata.Model = additionalMetadata.Model
		}
		if additionalMetadata.Config != nil {
			parsedPrompt.PromptMetadata.Config = additionalMetadata.Config
		}
	}
	return parsedPrompt
}

// RenderMetadata renders the metadata for the prompt.
func (dp *Dotprompt) RenderMetadata(source any, additionalMetadata *PromptMetadata) (PromptMetadata, error) {
	var parsedSource ParsedPrompt
	var err error
	switch v := source.(type) {
	case string:
		parsedSource, err = dp.Parse(v)
		if err != nil {
			return PromptMetadata{}, err
		}
	case ParsedPrompt:
		parsedSource = v
	default:
		return PromptMetadata{}, errors.New("invalid source type")
	}

	if additionalMetadata == nil {
		additionalMetadata = &PromptMetadata{}
	}
	selectedModel := additionalMetadata.Model
	if selectedModel == "" {
		selectedModel = parsedSource.PromptMetadata.Model
	}
	if selectedModel == "" {
		selectedModel = dp.defaultModel
	}

	modelConfig, ok := dp.modelConfigs[selectedModel].(map[string]any)
	if !ok {
		modelConfig = make(map[string]any)
	}
	metadata := []*PromptMetadata{}
	metadata = append(metadata, &parsedSource.PromptMetadata)
	metadata = append(metadata, additionalMetadata)

	return dp.ResolveMetadata(PromptMetadata{Config: modelConfig}, metadata)
}

// Assuming out and merges[i] are of type PromptMetadata
func mergeStructs(out, merge PromptMetadata) PromptMetadata {
	outVal := reflect.ValueOf(&out).Elem()
	mergeVal := reflect.ValueOf(merge)

	for i := 0; i < mergeVal.NumField(); i++ {
		field := mergeVal.Type().Field(i)
		value := mergeVal.Field(i)

		if !value.IsZero() {
			outVal.FieldByName(field.Name).Set(value)
		}
	}

	return out
}

// ResolveMetadata resolves and merges metadata.
func (dp *Dotprompt) ResolveMetadata(base PromptMetadata, merges []*PromptMetadata) (PromptMetadata, error) {
	out := base
	for _, merge := range merges {
		if merge == nil {
			continue
		}
		// config := out.Config
		// if config == nil {
		// 	config = make(map[string]any)
		// }
		out = mergeStructs(out, *merge)

		for key, value := range merge.Config {
			out.Config[key] = value
		}
	}
	out, err := dp.ResolveTools(out)
	if err != nil {
		return PromptMetadata{}, err
	}
	return dp.RenderPicoschema(out)
}

// ResolveTools resolves tools in the metadata.
func (dp *Dotprompt) ResolveTools(base PromptMetadata) (PromptMetadata, error) {
	out := base
	if out.Tools != nil {
		var outTools []string
		if out.ToolDefs == nil {
			out.ToolDefs = make([]ToolDefinition, 0)
		}

		for _, toolName := range out.Tools {
			if tool, exists := dp.tools[toolName]; exists {
				out.ToolDefs = append(out.ToolDefs, tool)
			} else if dp.toolResolver != nil {
				resolvedTool, err := dp.toolResolver(toolName)
				if err != nil {
					return PromptMetadata{}, err
				}
				if reflect.DeepEqual(resolvedTool, ToolDefinition{}) {
					return PromptMetadata{}, fmt.Errorf("Dotprompt: Unable to resolve tool '%s' to a recognized tool definition", toolName)
				}
				out.ToolDefs = append(out.ToolDefs, resolvedTool)
			} else {
				outTools = append(outTools, toolName)
			}
		}

		out.Tools = outTools
	}
	return out, nil
}

// RenderPicoschema renders the picoschema for the metadata.
func (dp *Dotprompt) RenderPicoschema(meta PromptMetadata) (PromptMetadata, error) {
	if meta.Output.Schema == nil && meta.Input.Schema == nil {
		return meta, nil
	}

	newMeta := meta
	if meta.Input.Schema != nil {
		schema, err := Picoschema(meta.Input.Schema, &PicoschemaOptions{
			SchemaResolver: func(name string) (JSONSchema, error) {
				return dp.WrappedSchemaResolver(name)
			},
		})
		if err != nil {
			return PromptMetadata{}, err
		}
		newMeta.Input.Schema = Schema(schema)
	}
	if meta.Output.Schema != nil {
		schema, err := Picoschema(meta.Output.Schema, &PicoschemaOptions{
			SchemaResolver: func(name string) (JSONSchema, error) {
				return dp.WrappedSchemaResolver(name)
			},
		})
		if err != nil {
			return PromptMetadata{}, err
		}
		newMeta.Output.Schema = Schema(schema)
	}
	return newMeta, nil
}

// WrappedSchemaResolver resolves schemas.
func (dp *Dotprompt) WrappedSchemaResolver(name string) (JSONSchema, error) {
	if schema, exists := dp.schemas[name]; exists {
		return schema, nil
	}
	if dp.schemaResolver != nil {
		return dp.schemaResolver(name)
	}
	return nil, nil
}

File: dart/dotprompt/lib/src/dotprompt.dart
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

/// Main Dotprompt class for parsing and rendering prompt templates.
///
/// This is the primary entry point for the Dotprompt library, providing
/// methods for parsing, compiling, and rendering prompt templates.
///
/// ## Key Features
///
/// - Template parsing with YAML frontmatter extraction
/// - Handlebars-style variable substitution
/// - Built-in helpers for common patterns
/// - Partial template support
/// - Tool and schema resolution
/// - Model configuration management
///
/// ## Example Usage
///
/// ```dart
/// final dotprompt = Dotprompt();
///
/// // Parse a template
/// final parsed = dotprompt.parse('''
/// ---
/// model: gemini-pro
/// ---
/// Hello {{name}}!
/// ''');
///
/// // Render with data
/// final result = await dotprompt.render(parsed.template, DataArgument(
///   input: {'name': 'World'},
/// ));
///
/// print(result.messages.first.content);  // "Hello World!"
/// ```
library;

import "dart:async";

import "package:handlebarrz/handlebarrz.dart";

import "error.dart";
import "helpers/media_helper.dart";
import "helpers/role_helper.dart";
import "models/models.dart";
import "parse.dart";
import "picoschema.dart";
import "store.dart";
import "types.dart";

/// Configuration options for a [Dotprompt] instance.
///
/// Allows customization of model defaults, helpers, partials, and resolvers.
class DotpromptOptions {
  /// Creates a new [DotpromptOptions].
  const DotpromptOptions({
    this.defaultModel,
    this.modelConfigs,
    this.helpers,
    this.partials,
    this.tools,
    this.schemas,
    this.partialResolver,
    this.toolResolver,
    this.schemaResolver,
    this.store,
  });

  /// The default model to use if none is specified in the template.
  final String? defaultModel;

  /// Model-specific configuration overrides.
  final Map<String, Map<String, dynamic>>? modelConfigs;

  /// Custom helper functions.
  final Map<String, Function>? helpers;

  /// Pre-registered partial templates.
  final Map<String, String>? partials;

  /// Pre-registered tool definitions.
  final Map<String, ToolDefinition>? tools;

  /// Pre-registered schemas (Picoschema or JSON Schema).
  final Map<String, Map<String, dynamic>>? schemas;

  /// Resolver for loading partial templates dynamically.
  final DotpromptPartialResolver? partialResolver;

  /// Resolver for loading tool definitions dynamically.
  final ToolResolver? toolResolver;

  /// Resolver for loading schemas dynamically.
  final SchemaResolver? schemaResolver;

  /// Store for loading prompts and partials.
  final PromptStore? store;
}

/// Main entry point for the Dotprompt library.
///
/// Provides methods for parsing, compiling, and rendering prompt templates.
class Dotprompt {
  /// Creates a new [Dotprompt] instance with optional configuration.
  Dotprompt([DotpromptOptions? options])
      : _options = options ?? const DotpromptOptions(),
        _partials = Map<String, String>.from(options?.partials ?? {}),
        _tools =
            Map<String, ToolDefinition>.from(options?.tools ?? {}),
        _schemas =
            Map<String, Map<String, dynamic>>.from(options?.schemas ?? {});

  final DotpromptOptions _options;
  final Map<String, String> _partials;
  final Map<String, ToolDefinition> _tools;
  final Map<String, Map<String, dynamic>> _schemas;

  /// The default model for this instance.
  String? get defaultModel => _options.defaultModel;

  /// Custom helpers registry.
  final Map<String, HelperFunction> _helpers = {};

  /// Defines a custom helper function.
  ///
  /// The helper will be available in templates as {{helperName ...}}.
  void defineHelper(String name, HelperFunction helper) {
    _helpers[name] = helper;
  }

  /// Defines a partial template.
  ///
  /// The partial will be available in templates as {{> partialName}}.
  void definePartial(String name, String template) {
    _partials[name] = template;
  }

  /// Defines a tool for use in prompts.
  void defineTool(ToolDefinition definition) {
    _tools[definition.name] = definition;
  }

  /// Defines a schema (Picoschema or JSON Schema).
  void defineSchema(String name, Map<String, dynamic> schema) {
    _schemas[name] = schema;
  }

  /// Parses a prompt template source into a [ParsedPrompt].
  ///
  /// Extracts the YAML frontmatter and template body.
  ///
  /// Throws [ParseException] if parsing fails.
  ParsedPrompt parse(String source) => Parser.parseDocument(source);

  /// Compiles a template for repeated rendering.
  ///
  /// Returns a function that can be called multiple times with different
  /// data to render the same template.
  Future<PromptFunction> compile(String source) async {
    final parsed = parse(source);
    await _resolvePartials(parsed.template);
    return _CompiledPromptFunction(this, parsed);
  }

  /// Renders a prompt template with the provided data.
  ///
  /// This is a convenience method that combines parsing and rendering.
  Future<RenderedPrompt> render(
    String source,
    DataArgument data, [
    Map<String, dynamic>? options,
  ]) async {
    final promptFn = await compile(source);
    return promptFn.render(data, options);
  }

  /// Renders the metadata for a template without rendering the full template.
  ///
  /// Useful when you need resolved metadata (tools, schemas) without the
  /// message content.
  Future<PromptMetadata> renderMetadata(String source) async {
    final parsed = parse(source);
    return _resolveMetadata(parsed.metadata);
  }

  /// Resolves partial references in a template.
  Future<void> _resolvePartials(String template) async {
    final partialPattern = RegExp(r"\{\{>\s*([a-zA-Z0-9_-]+)");
    final matches = partialPattern.allMatches(template);

    for (final match in matches) {
      final name = match.group(1)!;
      if (_partials.containsKey(name)) continue;

      // Try resolver first
      if (_options.partialResolver != null) {
        final content = await _options.partialResolver!(name);
        if (content != null) {
          _partials[name] = content;
          await _resolvePartials(content); // Recursive resolution
          continue;
        }
      }

      // Try store
      if (_options.store != null) {
        final data = await _options.store!.loadPartial(name, null);
        if (data != null) {
          _partials[name] = data.source;
          await _resolvePartials(data.source); // Recursive resolution
        }
      }
    }
  }

  /// Resolves metadata, including tools and schemas.
  Future<PromptMetadata> _resolveMetadata(PromptMetadata metadata) async {
    // Resolve model
    var model = metadata.model ?? _options.defaultModel;

    // Build config from model defaults and template config
    final config = <String, dynamic>{};
    if (model != null && _options.modelConfigs?.containsKey(model) == true) {
      config.addAll(_options.modelConfigs![model]!);
    }
    if (metadata.config != null) {
      config.addAll(metadata.config!);
    }

    // Resolve tools
    final toolDefs = <ToolDefinition>[];
    final unresolvedTools = <String>[];

    if (metadata.tools != null) {
      for (final toolName in metadata.tools!) {
        if (_tools.containsKey(toolName)) {
          toolDefs.add(_tools[toolName]!);
        } else if (_options.toolResolver != null) {
          final resolved = await _options.toolResolver!(toolName);
          if (resolved != null) {
            toolDefs.add(ToolDefinition.fromJson(resolved));
          } else {
            unresolvedTools.add(toolName);
          }
        } else {
          unresolvedTools.add(toolName);
        }
      }
    }

    // Process schemas (convert Picoschema to JSON Schema)
    InputConfig? input = metadata.input;
    OutputConfig? output = metadata.output;

    if (input?.schema != null && Picoschema.isPicoschema(input!.schema!)) {
      final jsonSchema = Picoschema.toJsonSchema(input.schema);
      input = InputConfig(
        schema: jsonSchema,
        defaultValues: input.defaultValues,
      );
    }

    if (output?.schema != null && Picoschema.isPicoschema(output!.schema!)) {
      final jsonSchema = Picoschema.toJsonSchema(output.schema);
      output = OutputConfig(
        format: output.format,
        schema: jsonSchema,
      );
    }

    return PromptMetadata(
      model: model,
      config: config.isNotEmpty ? config : null,
      input: input,
      output: output,
      tools: unresolvedTools.isNotEmpty ? unresolvedTools : null,
      toolDefs: toolDefs.isNotEmpty ? toolDefs : null,
      ext: metadata.ext,
      raw: metadata.raw,
    );
  }

  /// Renders a template with the given data.
  Future<RenderedPrompt> _renderInternal(
    ParsedPrompt parsed,
    DataArgument data,
    Map<String, dynamic>? options,
  ) async {
    // Build merged data context
    final mergedData = <String, dynamic>{};

    // Add input defaults
    if (parsed.input?.defaultValues != null) {
      mergedData.addAll(parsed.input!.defaultValues!);
    }

    // Add options input defaults
    if (options?["input"] case {"default": final Map defaults}) {
      mergedData.addAll(defaults.cast<String, dynamic>());
    }

    // Add provided input (overrides defaults)
    if (data.input != null) {
      mergedData.addAll(data.input!);
    }

    // Add context data with @ prefix for template access
    if (data.context != null) {
      for (final key in data.context!.keys) {
        mergedData[key] = data.context![key];
      }
    }

    // Render template using Handlebars
    final hb = Handlebars(escapeHtml: false);

    // Register built-in Dotprompt helpers
    hb.registerHelper("role", (args, options) {
      final role = args.isNotEmpty ? args[0].toString() : "user";
      return SafeString("<<<dotprompt:role:$role>>>");
    });

    hb.registerHelper("history", (args, options) {
      return SafeString("<<<dotprompt:history>>>");
    });

    hb.registerHelper("section", (args, options) {
      final name = args.isNotEmpty ? args[0].toString() : "default";
      return SafeString("<<<dotprompt:section $name>>>");
    });

    hb.registerHelper("media", (args, options) {
      final url = options.hash["url"]?.toString() ?? "";
      final contentType = options.hash["contentType"]?.toString() ?? "";
      return SafeString(
        "<<<dotprompt:media:url $url${contentType.isNotEmpty ? " $contentType" : ""}>>>",
      );
    });

    hb.registerHelper("json", (args, options) {
      if (args.isEmpty) return "";
      final value = args[0];
      if (value == null) return "null";
      if (value is String) return '"$value"';
      if (value is num || value is bool) return value.toString();
      return value.toString();
    });

    // Register ifEquals block helper
    hb.registerHelper("ifEquals", (args, options) {
      if (args.length >= 2 && args[0] == args[1]) {
        return options.fn(options.context);
      } else {
        return options.inverse(options.context);
      }
    });

    // Register unlessEquals block helper
    hb.registerHelper("unlessEquals", (args, options) {
      if (args.length >= 2 && args[0] != args[1]) {
        return options.fn(options.context);
      } else {
        return options.inverse(options.context);
      }
    });

    // Register custom user-defined helpers
    for (final entry in _helpers.entries) {
      hb.registerHelper(entry.key, entry.value);
    }

    // Register partials
    for (final entry in _partials.entries) {
      hb.registerPartial(entry.key, entry.value);
    }

    // Compile and render template
    final renderFn = hb.compile(parsed.template);
    final renderedString = renderFn(mergedData);

    // Convert rendered string to messages
    var messages = _toMessages(renderedString);

    // Merge in history messages if provided
    // History is inserted after any system messages from the template
    if (data.messages != null && data.messages!.isNotEmpty) {
      final historyMessages = data.messages!;
      final finalMessages = <Message>[];

      // Find the index after the last system message
      var insertIndex = 0;
      for (var i = 0; i < messages.length; i++) {
        if (messages[i].role == Role.system) {
          insertIndex = i + 1;
        }
      }

      // Insert system messages, then history, then rest
      finalMessages.addAll(messages.sublist(0, insertIndex));
      finalMessages.addAll(historyMessages);
      finalMessages.addAll(messages.sublist(insertIndex));
      messages = finalMessages;
    }

    // Build result config
    final resolvedMetadata = await _resolveMetadata(parsed.metadata);
    final resultConfig = resolvedMetadata.toConfig();

    // Add input config showing defaults that were used
    if (parsed.input?.defaultValues != null ||
        options?["input"]?["default"] != null) {
      resultConfig["input"] = {
        if (parsed.input?.defaultValues != null)
          "default": parsed.input!.defaultValues,
        if (options?["input"]?["default"] != null)
          "default": options!["input"]["default"],
      };
    }

    return RenderedPrompt(
      config: resultConfig,
      messages: messages,
    );
  }

  /// Converts rendered template string to messages.
  List<Message> _toMessages(String rendered) {
    final messages = <Message>[];
    var currentRole = Role.user; // Default role
    final currentParts = <Part>[];
    final currentText = StringBuffer();

    void flushText() {
      final text = currentText.toString();
      if (text.isNotEmpty) {
        // Check for media markers in text
        if (MediaHelper.containsMarker(text)) {
          final parts = _parseMediaMarkers(text);
          currentParts.addAll(parts);
        } else {
          currentParts.add(TextPart(text: text));
        }
        currentText.clear();
      }
    }

    void flushMessage() {
      flushText();
      if (currentParts.isNotEmpty) {
        messages.add(Message(role: currentRole, content: List.from(currentParts)));
        currentParts.clear();
      }
    }

    // Process the rendered string
    final lines = rendered.split("\n");
    for (var i = 0; i < lines.length; i++) {
      final line = lines[i];
      final isLastLine = i == lines.length - 1;

      // Check for role markers
      if (RoleHelper.containsMarker(line)) {
        final roleMatch = RegExp(
          "${RegExp.escape(RoleHelper.markerPrefix)}([^>]+)${RegExp.escape(RoleHelper.markerSuffix)}",
        ).firstMatch(line);
        if (roleMatch != null) {
          // Before switching roles, add trailing newline if not at start
          // (the newline that was between previous content and this role marker)
          if (currentText.isNotEmpty) {
            currentText.write("\n");
          }
          flushMessage();
          currentRole = Role.fromString(roleMatch.group(1)!);
          final remaining = line.replaceAll(roleMatch.group(0)!, "");
          // Don't trim - preserve leading/trailing whitespace in content
          if (remaining.isNotEmpty) {
            currentText.write(remaining);
          }
          continue;
        }
      }

      // Regular line
      if (currentText.isNotEmpty) {
        currentText.write("\n");
      }
      currentText.write(line);
    }

    flushMessage();

    // If no messages were created, create a single user message
    if (messages.isEmpty && rendered.trim().isNotEmpty) {
      return [
        Message(
          role: Role.user,
          content: [TextPart(text: rendered.trim())],
        ),
      ];
    }

    return messages;
  }

  /// Parses media markers in text into parts.
  List<Part> _parseMediaMarkers(String text) {
    final parts = <Part>[];
    final pattern = RegExp(
      "${RegExp.escape(MediaHelper.markerPrefix)}([^>]+)${RegExp.escape(MediaHelper.markerSuffix)}",
    );

    var lastEnd = 0;
    for (final match in pattern.allMatches(text)) {
      // Add text before marker
      if (match.start > lastEnd) {
        final textBefore = text.substring(lastEnd, match.start);
        if (textBefore.isNotEmpty) {
          parts.add(TextPart(text: textBefore));
        }
      }

      // Parse media marker
      final media = MediaHelper.extractMedia(match.group(0)!);
      if (media != null) {
        parts.add(MediaPart(media: media));
      }

      lastEnd = match.end;
    }

    // Add remaining text
    if (lastEnd < text.length) {
      final remaining = text.substring(lastEnd);
      if (remaining.isNotEmpty) {
        parts.add(TextPart(text: remaining));
      }
    }

    return parts;
  }
}

/// A compiled prompt function for repeated rendering.
abstract interface class PromptFunction {
  /// Renders the prompt with the given data.
  Future<RenderedPrompt> render(
    DataArgument data, [
    Map<String, dynamic>? options,
  ]);

  /// Gets the parsed prompt.
  ParsedPrompt get prompt;
}

/// Internal implementation of [PromptFunction].
class _CompiledPromptFunction implements PromptFunction {
  _CompiledPromptFunction(this._dotprompt, this._parsed);

  final Dotprompt _dotprompt;
  final ParsedPrompt _parsed;

  @override
  Future<RenderedPrompt> render(
    DataArgument data, [
    Map<String, dynamic>? options,
  ]) =>
      _dotprompt._renderInternal(_parsed, data, options);

  @override
  ParsedPrompt get prompt => _parsed;
}
